In [1]:
%autosave 2

Autosaving every 2 seconds


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
# Other imports
from pprint import pprint

# ~Let's loaded the data and explore a bit

In [4]:
df_train = pd.read_csv('dataset/train.csv')
df_train.head(5)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [5]:
df_train.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In [6]:
df_train.shape  # output as (#Row, #Column)

(1460, 81)

So, there are **1460 rows** and **81 columns** in the training dataset.

In [7]:
# df_train.info()

In [8]:
df_train.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

- **Now we want to know about the Missing value columns**

In [9]:
missaing_value_columns = {
    column: df_train[column].isnull().sum(axis=0) 
    for column in df_train.columns if df_train[column].isnull().any()
}
missaing_value_columns  # Missaing Value Columns & their Count

{'Alley': 1369,
 'BsmtCond': 37,
 'BsmtExposure': 38,
 'BsmtFinType1': 37,
 'BsmtFinType2': 38,
 'BsmtQual': 37,
 'Electrical': 1,
 'Fence': 1179,
 'FireplaceQu': 690,
 'GarageCond': 81,
 'GarageFinish': 81,
 'GarageQual': 81,
 'GarageType': 81,
 'GarageYrBlt': 81,
 'LotFrontage': 259,
 'MasVnrArea': 8,
 'MasVnrType': 8,
 'MiscFeature': 1406,
 'PoolQC': 1453}

# ~Knowing About the dataset column
1. Fatures/Otput 
2. Column Types 
3. Relevance, Nature and Relation with Output

# 1. Features/Output 
**At first, we have to identify the input features and the output column. Form the challenge description and dataset description, we can see that,**
1. Output: SalePrice column
2. Features: All the other columns

# 2. Column Types
In Machine Learning datasets, it is well known matter that, dataset contains two types of column:
1. **Numerical Value Column**: Values are numbers. Value can either be integer or floating point.
2. **Categorical Value Column**: Values are categories. This is represented using either string or number (integers in common).


So, in the very beginning, we need to identify which columns posses **Categorical Value** and which does **Numerical Value**. This can be identified from the **dataset description**.

#### Some Steps can be taken to find out the categorical/numerical columns:
1. Column values having string can be categorical value, if the column is relevant to the prblem we are solving
2. Columns having Numeric values can be categorical. The rating of house can be 1, 2, ..., 10. For example, OverallQual, OverallCond in this challenge dataset <br/>
- All the other columns are **Numerical Value Columns**

### Catrgorical Columns
- Categorical values have been identified and listed in an Excel file named **dataset_observation.csv**

In [10]:
categorical_columns = pd.read_csv('dataset_observation.csv')
categorical_columns = categorical_columns['categorical_columns'].tolist()
print('No. of categorical value columns:', len(categorical_columns), '\n')
print(categorical_columns)

No. of categorical value columns: 47 

['MSSubClass', 'MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'OverallQual', 'OverallCond', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Heating', 'CentralAir', 'Electrical', 'KitchenQual', 'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature', 'MoSold', 'YrSold', 'SaleType', 'SaleCondition']


**Now, we'll find the categorical columns which have missing values**

In [11]:
categorical_columns_with_missing_value = list(set(categorical_columns) & set(missaing_value_columns.keys()))
categorical_columns_with_missing_value = {
    column: missaing_value_columns[column] for column in categorical_columns_with_missing_value
}
print(
    'No. of categorical value columns having missing value:', len(categorical_columns_with_missing_value), '\n'
)
pprint(categorical_columns_with_missing_value)

No. of categorical value columns having missing value: 16 

{'Alley': 1369,
 'BsmtCond': 37,
 'BsmtExposure': 38,
 'BsmtFinType1': 37,
 'BsmtFinType2': 38,
 'BsmtQual': 37,
 'Electrical': 1,
 'Fence': 1179,
 'FireplaceQu': 690,
 'GarageCond': 81,
 'GarageFinish': 81,
 'GarageQual': 81,
 'GarageType': 81,
 'MasVnrType': 8,
 'MiscFeature': 1406,
 'PoolQC': 1453}


We can see that, most of the data in columns Alley, MiscFeature, PoolQC, Fence are missing. So, we can ignore these columns.

### Numerical Columns

In [12]:
numerical_columns = list(set(df_train)-set(categorical_columns))
print('No. of numerical value columns:', len(numerical_columns), '\n')
print(numerical_columns)

No. of numerical value columns: 34 

['GrLivArea', 'BsmtFinSF1', 'BsmtUnfSF', 'FullBath', 'BsmtFinSF2', 'TotRmsAbvGrd', 'LotFrontage', '3SsnPorch', 'EnclosedPorch', 'MasVnrArea', 'GarageCars', 'ScreenPorch', 'BsmtFullBath', 'HeatingQC', 'PoolArea', 'GarageYrBlt', '2ndFlrSF', 'HalfBath', 'Fireplaces', 'LowQualFinSF', '1stFlrSF', 'BedroomAbvGr', 'WoodDeckSF', 'SalePrice', 'GarageArea', 'YearBuilt', 'YearRemodAdd', 'BsmtHalfBath', 'OpenPorchSF', 'MiscVal', 'TotalBsmtSF', 'KitchenAbvGr', 'Id', 'LotArea']


**Now, we'll find the numerical columns which have missing values**

In [13]:
numerical_columns_with_missing_value = list(set(numerical_columns) & set(missaing_value_columns.keys()))
numerical_columns_with_missing_value = {
    column: missaing_value_columns[column] for column in numerical_columns_with_missing_value
}
print(
    'No. of numerical value columns having missing value:', len(numerical_columns_with_missing_value), '\n'
)
pprint(numerical_columns_with_missing_value)

No. of numerical value columns having missing value: 3 

{'GarageYrBlt': 81, 'LotFrontage': 259, 'MasVnrArea': 8}


Only 3 variables have missing values. Whooooooaaaa !!!!

# 3. Relevance, Nature and Relation with Output